### PR TITLE
Tagging of floats and suspend/resume

### PR DESCRIPTION
**Pull request by u-fischer**
This changes the implementation of float and caption tagging to sockets and handles floats in multicol (issue https://github.com/latex3/tagging-project/issues/705).  The tagging of captions has been slightly changed. 

Additionally it adds l3layer version of \SuspendTagging and \ResumeTagging and uses them (tagpdf has been updated for this).

## Status of pull request

- Ready to merge

## Checklist of required changes before merge will be approved
- [x] Test file(s) added
- [x] Version and date string updated in changed source files
- [n/a] Relevant `\changes` entries in source included
- [x] Relevant `changes.txt` updated
- [n/a] Rollback provided (if necessary)?
- [n/a] `ltnewsX.tex` (and/or `latexchanges.tex`) updated

### CODE REVIEW

File: required/latex-lab/latex-lab-float.dtx [hunk 1]
```diff
@@ -383,7 +462,8 @@
 %    \end{macrocode}
 % This starts the structure for the float.
 %    \begin{macrocode}
-  \@nameuse{@@_float_begin:}%
+  \csname @@_float_init:\endcsname
```
Comment (josephwright) [inline required/latex-lab/latex-lab-float.dtx:465]: Why this change?

**Review by josephwright — CHANGES_REQUESTED** [review]
Minor comments only

File: required/latex-lab/latex-lab-float.dtx (hunk 1 above)
Comment (u-fischer) [reply required/latex-lab/latex-lab-float.dtx:465]: The `\@@_float_begin:` command has been replaced by `\UseTaggingSocket{float/begin}`. The `\@@_float_init:`  is new, it takes care of the bits that must be initialized outside a tagging socket (like a counter for link targets). Why I used `\csname/\endcsname` instead of `\@nameuse`  I don't know ;-). Do we have a preference here?

**Review by FrankMittelbach — CHANGES_REQUESTED** [review]
mainly a bunch of questions and comments, perhaps not much that needs "changing" but ...

File: required/latex-lab/latex-lab-float.dtx (hunk 1 above)
Comment (u-fischer) [reply required/latex-lab/latex-lab-float.dtx:465]: yes, probably. But I'm not sure yet if the command itself should be public, or if there should be a wrapper socket or similar. So I would keep that on hold for now.

File: base/lttagging.dtx [hunk 2]
```diff
@@ -261,6 +244,13 @@
 \ExplSyntaxOff 
 %    \end{macrocode}
 % \end{plugdecl}
+% 
+% \begin{socketdecl}{tagsupport/para/begin,tagsupport/para/end}
+% These sockets are currently defined in tagpdf. They overwrite
+% definitions in the latex-lab-block code. There is also a simpler
+% definition that probably should be a general socket too.
+% TODO: move this into lttagging.
```
[the remark this answers is not in the document]
Comment (u-fischer) [reply base/lttagging.dtx:252]: this is currenlty only an empty declaration with a comment, the actual socket declaration are as written in tagpdf but they should move here. I will prepare that for the next tagpdf update.

File: required/latex-lab/latex-lab-math.dtx [hunk 3]
```diff
@@ -2470,7 +2470,7 @@
 %    \end{macrocode}
 %    Also for "false?"
 %    \begin{macrocode}
-\def\measuring@true{\let\ifmeasuring@\iftrue\tag_stop:}
+\def\measuring@true{\let\ifmeasuring@\iftrue\tag_suspend:n{\measuring}}
```
[the remark this answers is not in the document]
Comment (u-fischer) [reply required/latex-lab/latex-lab-math.dtx:2473]: I thought one \measuring is probably indication enough that we are in a box and that we could save the resume here.

File: required/latex-lab/latex-lab-table.dtx [hunk 4]
```diff
@@ -1425,9 +1425,9 @@
       }
     \hbox to\z@{\hss\parbox[t]\LTcapwidth{%
     \reset@font
-    \tag_stop:n{caption}
+    \tag_suspend:n{caption}
     \sbox\@tempboxa{#1{#2:~}#3}%
-    \tag_start:n{caption}
+    \tag_resume:n{caption}
```
[the remark this answers is not in the document]
Comment (u-fischer) [reply required/latex-lab/latex-lab-table.dtx:1430]: yes, if you save a box and then reuse it directly then you want its internal tagging to be preserved. But it could be useful to have an `\untaggedsbox` command.

File: required/latex-lab/latex-lab-float.dtx [hunk 5]
```diff
@@ -260,15 +283,71 @@
...
+%    \end{macrocode}
+% \end{plugdecl}
+%
+% \begin{plugdecl}{default  (tagsupport/float/hmode/end)}
+% This plug should be used if a float is called in hmode
+% and the end of the float it then restarts the MC.
+%    \begin{macrocode}
+\NewSocketPlug{tagsupport/float/hmode/end}{default}
+  {
+    \@@_float_start_par:
```
[the remark this answers is not in the document]
Comment (u-fischer) [reply required/latex-lab/latex-lab-float.dtx:325]: perhaps, in the next round after looking at more float packages ...

File: required/latex-lab/latex-lab-float.dtx [hunk 6]
```diff
@@ -453,10 +533,112 @@
...
+ }
+%    \end{macrocode}
+%
+% \begin{socketdecl}{caption/label}
+% This socket is a lightweight start for 
+% some interface to format the label or add a font command. 
+% The argument is the label text. 
+% The default plug \texttt{kernel} adds a colon
+% and a space.
+% TODO: revisit.
```
[the remark this answers is not in the document]
Comment (u-fischer) [reply required/latex-lab/latex-lab-float.dtx:564]: yes, and all caption packages ;-)

File: required/latex-lab/latex-lab-float.dtx (hunk 1 above)
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-float.dtx:465]: ok, but then I would add a TODO  there

File: required/latex-lab/latex-lab-math.dtx (hunk 3 above)
[the remark this answers is not in the document]
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-math.dtx:2473]: I would put it in in the right place, even if commented out for now, so that it it is lear where to uncomment it, but up to you.

File: required/latex-lab/latex-lab-table.dtx (hunk 4 above)
[the remark this answers is not in the document]
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-table.dtx:1430]: I was thinking the other way around, make sbox untagged by default and have a \taggedsbox, but it kind of depends what happens more often.

File: required/latex-lab/latex-lab-float.dtx (hunk 5 above)
[the remark this answers is not in the document]
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-float.dtx:325]: ok

File: required/latex-lab/latex-lab-float.dtx (hunk 6 above)
[the remark this answers is not in the document]
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-float.dtx:564]: but then perhaps the TODO should say precisely that :-)

File: required/latex-lab/latex-lab-math.dtx (hunk 3 above)
[the remark this answers is not in the document]
Comment (u-fischer) [reply required/latex-lab/latex-lab-math.dtx:2473]: well actually "the right place" is not trivial, as naturally most is in amsmath directly. So for now I would leave that as it is.

[the remark this answers is not in the document]
Comment (FrankMittelbach) [reply required/latex-lab/latex-lab-math.dtx:2473]: but shouldn't all of that be moved to amsmath anyway, eventually? rather than redefining `\measuring@true` here. Okay, doesn't have to happen now, but eventually ...

**Review by FrankMittelbach — APPROVED** [review]
